browser/search_engines/android/java/src/org/chromium/brave/browser/search_engines/settings/CustomSearchEnginesUtil.java

```diff
@@ -0,0 +1,173 @@
+import org.chromium.chrome.browser.profiles.Profile;
+import org.chromium.chrome.browser.search_engines.TemplateUrlServiceFactory;
+import org.chromium.chrome.browser.ui.favicon.FaviconUtils;
+import org.chromium.components.favicon.LargeIconBridge;
+import org.chromium.components.favicon.LargeIconBridge.GoogleFaviconServerCallback;
+import org.chromium.components.favicon.LargeIconBridge.LargeIconCallback;
+import org.chromium.components.search_engines.TemplateUrlService;
+import org.chromium.net.NetworkTrafficAnnotationTag;
+import org.chromium.url.GURL;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.LinkedList;
+import java.util.List;
+
+public class CustomSearchEnginesUtil {
+
+    private static final String TAG = "CustomSearchEngines";
+
+    private static final String CUSTOM_SEARCH_ENGINES = "custom_search_engines";
+
+    public static String KEYWORD = "keyword";
+
+    private static void saveCustomSearchEngines(List<String> customSearchEnginesList) {
+        if (customSearchEnginesList == null) {
+            return;
+        }
+
+        try {
+            JSONArray jsonArray = new JSONArray(customSearchEnginesList);
+            ChromeSharedPreferences.getInstance()
+                    .writeString(CUSTOM_SEARCH_ENGINES, jsonArray.toString());
+        } catch (Exception e) {
+            Log.e(TAG, "Error saving search engines", e);
+        }
+    }
+
+    public static void addCustomSearchEngine(String searchEngineKeyword) {
+        List<String> customSearchEnginesList = getCustomSearchEngines();
+        if (customSearchEnginesList.isEmpty()) {
+            customSearchEnginesList = new LinkedList<>();
+        }
+        if (!customSearchEnginesList.contains(searchEngineKeyword)) {
+            customSearchEnginesList.add(searchEngineKeyword);
+            saveCustomSearchEngines(customSearchEnginesList);
+        }
+    }
+
+    public static List<String> getCustomSearchEngines() {
+        List<String> customSearchEnginesList = new LinkedList();
+        String savedSearchEngines =
+                ChromeSharedPreferences.getInstance().readString(CUSTOM_SEARCH_ENGINES, null);
+
+        if (savedSearchEngines == null) {
+            return customSearchEnginesList;
+        }
+
+        try {
+            // Wrap the array in an object to make it valid JSON
+            JSONObject wrapper =
+                    new JSONObject("{\"customSearchEngines\":" + savedSearchEngines + "}");
+            JSONArray searchEnginesArray = wrapper.getJSONArray("customSearchEngines");
+
+            for (int i = 0; i < searchEnginesArray.length(); i++) {
+                customSearchEnginesList.add(searchEnginesArray.getString(i));
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return customSearchEnginesList;
+    }
+
+    public static boolean isCustomSearchEngineAdded(String searchEngineKeyword) {
```

Comment:
why are we checking some custom pref instead of pulling from the template url service?

Reply:
We also save user added search engines in prefs. so instead of checking directly with native, i am checking for the pref if it exists.

Reply:
Why would you duplicate it in a pref?

Reply:
@bridiver Currently, when a user adds a new search engine, it doesn't appear in the main "Search Engines" list right away because we filter out non-prepopulated search engines. It only becomes visible after the user sets it as the default.

However, we do save the search engine preferences upon adding. So when the user later visits the "Search Engines" page, we load any user-added search engines from the saved preferences and include them in the list.